src/applets/budgie-menu/backend/Tracker.vala

```diff
@@ -0,0 +1,225 @@
+
+	private AppInfoMonitor monitor;
+	private uint timeout_id = 0;
+
+	/**
+	 * Signal emitted whenever a change to the application state
+	 * occurs.
+	 */
+	public signal void changed();
+
+	public Tracker() {
+		Object();
+	}
+
+	construct {
+		this.categories = new Gee.ArrayList<Category>();
+
+		this.monitor = AppInfoMonitor.@get();
+		this.monitor.changed.connect(() => {
+			this.queue_refresh();
+		});
+
+		// Start building the tree right now
+		this.refresh();
+	}
+
+	/**
+	 * Queue an update of the application system to run.
+	 *
+	 * The time to wait before refreshing can be set by passing
+	 * in the number of seconds. By default the time is 3 seconds.
+	 */
+	public void queue_refresh(int seconds = 3) {
+		// Reset the refresh timer if an update is already queued
+		if (this.timeout_id != 0) {
+			Source.remove(this.timeout_id);
+			this.timeout_id = 0;
+		}
+
+		// Update the application system after the timeout
+		this.timeout_id = Timeout.add(seconds, () => {
+			this.refresh();
+			this.timeout_id = 0;
+			return Source.REMOVE;
+		});
+	}
+
+	/**
+	 * Rebuild the entire app and category indexes.
+	 *
+	 * This iterates over all AppInfos on the system, so it is likely to be
+	 * costly to call this function.
+	 */
+	private void refresh() {
+		this.categories.clear();
+
+		/*
+		 * Add our categories, adhearing to the Freedesktop Menu spec.
+		 *
+		 * Inclusions and exclusions are sourced from multiple places,
+		 * including the UbuntuBudgie fork of applications-menu, gnome-menus,
+		 * the Freedeskop Menus spec, and by ourselves.
+		 */
+
+		this.categories.add(new Category(_("Accessories")) {
+			included_categories = { "Utility" },
+			/*
+			 * The spec states that Accessibility must have either the Utility or Settings categories,
+			 * and we have a separate accessibility category, so don't put those applications here.
+			 */
+			excluded_categories = { "Accessibility", "System", "X-GNOME-Utilities" },
+			excluded_applications = {
+				"eog.desktop",
+				"gucharmap.desktop",
+				"org.gnome.DejaDup.desktop",
+				"org.gnome.Dictionary.desktop",
+				"org.gnome.DiskUtility.desktop",
+				"org.gnome.Evince.desktop",
+				"org.gnome.FileRoller.desktop",
+				"org.gnome.font-viewer.desktop",
+				"org.gnome.Screenshot.desktop",
+				"org.gnome.seahorse.Application.desktop",
+				"org.gnome.Terminal.desktop",
+				"org.gnome.tweaks.desktop",
+				"org.gnome.Usage.desktop",
+				"plank.desktop",
+				"simple-scan.desktop",
+				"vinagre.desktop",
+				"yelp.desktop"
+			}
+		});
+
+		this.categories.add(new Category(_("Education")) {
+			included_categories = { "Education" },
+			excluded_categories = { "Science" }
+		});
+
+		this.categories.add(new Category(_("Games")) {
+			included_categories = { "Game" }
+		});
+
+		this.categories.add(new Category(_("Graphics")) {
+			included_categories = { "Graphics" },
+			excluded_applications = {
+				"eog.desktop",
```

Comment:
EOG should not be excluded. It deals with graphics (images) and is currently listed under Graphics.